graph/src/data/schema.rs

```diff
@@ -606,6 +714,192 @@ impl Schema {
+            .into_iter()
+            .map(|(name, _)| name.clone())
+            .collect::<Vec<String>>();
+        let mut imported_types = self
+            .imported_types()
+            .into_iter()
+            .map(|(imported_type, _)| match imported_type {
+                ImportedType::Name(name) => name.clone(),
+                ImportedType::NameAs(_, az) => az.clone(),
+            })
+            .collect::<Vec<String>>();
+        all_types.append(&mut imported_types);
+
+        if let Some(_) = all_types.iter().find(|typ| typ.eq(&name)) {
+            return vec![SchemaValidationError::FulltextNameOverlap(name.to_string())];
+        } else {
+            return vec![];
+        }
+    }
+
+    fn validate_fulltext_directive_language(
+        &self,
+        fulltext: &Directive,
+    ) -> Vec<SchemaValidationError> {
+        let language_value = fulltext
+            .arguments
+            .iter()
+            .find(|(key, _)| key.eq("language"))
+            .map(|(_, value)| value);
+        let language = match language_value {
+            Some(Value::Enum(language)) => language,
+            _ => return vec![SchemaValidationError::FulltextLanguageUndefined],
+        };
+        match FulltextLanguage::try_from(language) {
+            Ok(_) => vec![],
+            Err(_) => vec![SchemaValidationError::FulltextLanguageInvalid(
+                language.to_string(),
+            )],
+        }
+    }
+
+    fn validate_fulltext_directive_algorithm(
+        &self,
+        fulltext: &Directive,
+    ) -> Vec<SchemaValidationError> {
+        let algorithm_value = fulltext
+            .arguments
+            .iter()
+            .find(|(key, _)| key.eq("algorithm"))
+            .map(|(_, value)| value);
+        let algorithm = match algorithm_value {
+            Some(Value::Enum(algorithm)) => algorithm,
+            _ => return vec![SchemaValidationError::FulltextAlgorithmUndefined],
+        };
+        match FulltextAlgorithm::try_from(algorithm) {
+            Ok(_) => vec![],
+            Err(_) => vec![SchemaValidationError::FulltextAlgorithmInvalid(
+                algorithm.to_string(),
+            )],
+        }
+    }
+
+    fn validate_fulltext_directive_includes(
+        &self,
+        fulltext: &Directive,
+    ) -> Vec<SchemaValidationError> {
+        // Only allow fulltext directive on local types
+        let local_types: Vec<&ObjectType> = self
+            .document
+            .get_object_type_definitions()
+            .into_iter()
+            .collect();
+
+        // Validate that each entity in fulltext.include exists
+        let include_value = fulltext
+            .arguments
+            .iter()
+            .find(|(key, _)| key.eq("include"))
+            .map(|(_, value)| value);
+        let includes = match include_value {
+            Some(Value::List(includes)) if includes.len() > 0 => includes,
+            _ => return vec![SchemaValidationError::FulltextIncludeUndefined],
```

Comment:
"Undefined" is misleading here as it _is_ may be defined. "invalid" or "not a list" would be more appropriate for cases where there is a value but it's not a list.